libcxx/include/support/win32/atomic_msvc.h

```diff
@@ -76,25 +93,47 @@ void __atomic_load(const volatile _Type *__ptr, _Type *__ret, int __memorder) {
 }
 
 template<class _Type, detail::_enable_if_sized_as<_Type,1> = 0>
-void __atomic_store_relaxed(const volatile _Type *__ptr, _Type *__val) {
-    __int8 __tmp = reinterpret_cast<__int8&>(*__val);
-    __iso_volatile_store8((volatile __int8 *)__ptr, __tmp);
+void __atomic_store_relaxed(volatile _Type *__ptr, _Type *__val) {
+    auto __t = reinterpret_cast<__int8 *>(__val);
+    auto __d = reinterpret_cast<volatile __int8 *>(__ptr);
+#ifdef _LIBCUDACXX_MSVC_HAS_NO_ISO_INTRIN
+    (void)_InterlockedExchange8(__d, *__t);
+#else
+    __iso_volatile_store8(__d, __t);
+#endif
 }
 template<class _Type, detail::_enable_if_sized_as<_Type,2> = 0>
-void __atomic_store_relaxed(const volatile _Type *__ptr, _Type *__val) {
-    __int16 __tmp = reinterpret_cast<__int16&>(*__val);
-    __iso_volatile_store16((volatile __int16 *)__ptr, __tmp);
+void __atomic_store_relaxed(volatile _Type *__ptr, _Type *__val) {
+    auto __t = reinterpret_cast<__int16 *>(__val);
+    auto __d = reinterpret_cast<volatile __int16 *>(__ptr);
+#ifdef _LIBCUDACXX_MSVC_HAS_NO_ISO_INTRIN
+    (void)_InterlockedExchange16(__d, *__t);
+#else
+    __iso_volatile_store16(__d, __t);
+#endif
 }
 template<class _Type, detail::_enable_if_sized_as<_Type,4> = 0>
-void __atomic_store_relaxed(const volatile _Type *__ptr, _Type *__val) {
-    __int32 __tmp = reinterpret_cast<__int32&>(*__val);
-    __iso_volatile_store32((volatile __int32 *)__ptr, __tmp);
+void __atomic_store_relaxed(volatile _Type *__ptr, _Type *__val) {
+    auto __t = reinterpret_cast<__int32 *>(__val);
+    auto __d = reinterpret_cast<volatile __int32 *>(__ptr);
+#ifdef _LIBCUDACXX_MSVC_HAS_NO_ISO_INTRIN
+    // int cannot be converted to long?...
+    (void)_InterlockedExchange(reinterpret_cast<volatile long *>(__d), *__t);
```

Comment:
MSVC wouldn't accept an implicit cast to `long`.

I don't think we need any explicit guarantees that long is always going to be 32 bits on MSVC.